doctr/models/recognition/parseq/pytorch.py

```diff
@@ -0,0 +1,493 @@
+        # with small modifications
+
+        max_num_chars = int(seqlen.max().item())  # get longest sequence length in batch
+        perms = [torch.arange(max_num_chars, device=seqlen.device)]
+
+        max_perms = math.factorial(max_num_chars) // 2
+        num_gen_perms = min(3, max_perms)
+        if max_num_chars < 5:
+            # Pool of permutations to sample from. We only need the first half (if complementary option is selected)
+            # Special handling for max_num_chars == 4 which correctly divides the pool into the flipped halves
+            if max_num_chars == 4:
+                selector = [0, 3, 4, 6, 9, 10, 12, 16, 17, 18, 19, 21]
+            else:
+                selector = list(range(max_perms))
+            perm_pool = torch.as_tensor(list(permutations(range(max_num_chars), max_num_chars)), device=seqlen.device)[
+                selector
+            ]
+            # If the forward permutation is always selected, no need to add it to the pool for sampling
+            perm_pool = perm_pool[1:]
+            final_perms = torch.stack(perms)
+            if len(perm_pool):
+                i = self.rng.choice(len(perm_pool), size=num_gen_perms - len(final_perms), replace=False)
+                final_perms = torch.cat([final_perms, perm_pool[i]])  # type: ignore[index]
+        else:
+            perms.extend(
+                [torch.randperm(max_num_chars, device=seqlen.device) for _ in range(num_gen_perms - len(perms))]
+            )
+            final_perms = torch.stack(perms)
+
+        comp = final_perms.flip(-1)
+        final_perms = torch.stack([final_perms, comp]).transpose(0, 1).reshape(-1, max_num_chars)
+
+        sos_idx = torch.zeros(len(final_perms), 1, device=seqlen.device)
+        eos_idx = torch.full((len(final_perms), 1), max_num_chars + 1, device=seqlen.device)
+        combined = torch.cat([sos_idx, final_perms + 1, eos_idx], dim=1).int()
+        if len(combined) > 1:
+            combined[1, 1:] = max_num_chars + 1 - torch.arange(max_num_chars + 1, device=seqlen.device)
+        # we pad to max length with eos idx to fit the mask generation
+        return F.pad(
+            combined, (0, self.max_length + 1 - combined.shape[-1]), value=max_num_chars + 2
+        )  # (num_perms, self.max_length + 1)
+
+    def generate_permutations_attention_masks(self, permutation: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        # Generate source and target mask for the decoder attention.
+        sz = permutation.shape[0]
+        mask = torch.ones((sz, sz), device=permutation.device)
+
+        for i in range(sz):
+            query_idx = permutation[i]
+            masked_keys = permutation[i + 1 :]
+            mask[query_idx, masked_keys] = 0.0
+        source_mask = mask[:-1, :-1].clone()
+        mask[torch.eye(sz, dtype=torch.bool, device=permutation.device)] = 0.0
+        target_mask = mask[1:, :-1]
```

Comment:
So we're using `0` and `1` for mask and authors used `float(-inf)` and `0` respectively, I'm a correct ?

Reply:
Yep for our MHA implementation 0 is masked (transformer decoder can't "see" it we replace it inside scaled dot product to -inf this masking is needed for Transformer decoder otherwise the model would be able to "cheat") and 1 is visible the softmax activation does the rest :)

To overcome the question why we don't set it directly to -inf
Short answer: this would raise problems on ONNX exporting :)